src/critics/obstacles_critic.cpp

```diff
@@ -22,29 +20,40 @@ void ObstaclesCritic::initialize()
     power_, weight_, consider_footprint_ ? "footprint" : "circular");
 }
 
-void ObstaclesCritic::score(
-  const geometry_msgs::msg::PoseStamped & /*robot_pose*/,
-  const models::State & /*state*/, const xt::xtensor<double, 3> & trajectories,
-  const xt::xtensor<double, 2> & /*path*/, xt::xtensor<double, 1> & costs,
-  nav2_core::GoalChecker * /*goal_checker*/)
+void ObstaclesCritic::score(models::CriticFunctionData & data)
 {
-  for (size_t i = 0; i < trajectories.shape()[0]; ++i) {
+
+  if (!enabled_) {
+    return;
+  }
+
+  bool all_trajectories_collide = true;
+  for (size_t i = 0; i < data.trajectories.shape(0); ++i) {
     bool trajectory_collide = false;
 
     unsigned char trajectory_cost = nav2_costmap_2d::FREE_SPACE;
-    for (size_t j = 0; j < trajectories.shape()[1]; ++j) {
+    for (size_t j = 0; j < data.trajectories.shape(1); ++j) {
       unsigned char pose_cost = costAtPose(
-        trajectories(i, j, 0), trajectories(i, j, 1), trajectories(i, j, 2));
+        data.trajectories(i, j, 0), data.trajectories(i, j, 1), data.trajectories(i, j, 2));
       trajectory_cost = std::max(trajectory_cost, pose_cost);
 
-      if (inCollision(pose_cost)) {
+      if (inCollision(trajectory_cost)) {
         trajectory_collide = true;
         break;
       }
+
+      if (j == data.trajectories.shape(1) - 1) {
+        all_trajectories_collide = false;
+      }
     }
 
-    costs[i] +=
-      trajectory_collide ? collision_cost_ : scoreCost(trajectory_cost);
+    data.costs[i] += trajectory_collide ? collision_cost_ : scoreCost(trajectory_cost);
+  }
+
+
+  if (all_trajectories_collide) {
+    data.stop_flag = true;
+    RCLCPP_WARN(logger_, "All trajectories collide! Reset Optimizer");
```

Comment:
What does resetting optimizer mean? Wouldn't this fail the iteration and if we fail all iterations we fail the control request (e.g. throw exception) so that we can execute recovery behaviors?

Reply:
Reset made all state velocities and control sequence equal zero, that means we start searching possible new trajectories from 0. We could easily find some back motions, or resample whole batch from absolutely another initial conditions to find more safe ahead motion

Reply:
> Current behavior is very good, but you mentioned that there were times that all trajectories collided but before this change it would just pass back to the controller server without notifying anyone of that fact. How often was that happening?

before recent changes in art branch it was often. Now it is rare, but sometimes happen.

 For the situation when the robot had something get in front of it suddenly and then it would stop, back up, and move around it -- would this be triggered, or was that all totally-functioning behavior that didn't have all_trajectories_collide
 
 I just added this feature, so before it wasn't caused by this, optimizer just reoptimized trajectory for such behavior. Using reset made this behavior even better. But now i'm not sure we must throw and run recover coz optimizer mostly after resetting could find feasible trajectory

Reply:
But the question is - what control should we pass in this case... Stop robot, or maybe just some portion of last control, to make change more smooth ?

Reply:
> before recent changes in art branch it was often. Now it is rare, but sometimes happen.

In what situations was it happening before?

> I just added this feature, so before it wasn't caused by this, optimizer just reoptimized trajectory for such behavior. 

I understand, but now that its been added, would that back-out situation trigger the `reset()` and `throw` to say that we failed to compute a trajectory? Or does that do it itself without any such failure?

If that back-out situation I provided above doesn't actually trigger a failure, then what situations does, causing you to say "before recent change in art branch it was often"?

I'm trying to understand that since this was silently failing before, what times it was actually failing that we will now actually be notified about.

> Using reset made this behavior even better. But now i'm not sure we must throw and run recovery behavior coz optimizer mostly after resetting could find feasible trajectory

If we reset and it works then, I agree. But if we reset and it doesn't, then we need to notify / throw of a failure. I suggest:
- We try, if we fail, reset and try again all in 1 cycle.
- If at that point we fail, then we throw. 

That way we silently fail 1 time to try to see if we can recover with a reset. If we cannot, then we should throw a failure message

> But the question is - what control should we pass in this case... Stop robot, or maybe just some portion of last control, to make change more smooth ?

If we cannot find a control in a given cycle, throw an exception. [The controller server will catch it](https://github.com/ros-planning/navigation2/blob/main/nav2_controller/src/controller_server.cpp#L460-L486) and retry within a buffered window to see if the controller can succeed. If not, it escillates into a recovery behavior. By default, this is disabled, but we could change that, I think there should be some small tolerance.

Reply:
You could implement that by having the controller have:

```
try {
  geometry_msgs::msg::TwistStamped cmd =
    optimizer_.evalControl(robot_pose, robot_speed, transformed_plan, goal_checker);
} catch (const std::runtime_error & e) {
  geometry_msgs::msg::TwistStamped cmd =
    optimizer_.evalControl(robot_pose, robot_speed, transformed_plan, goal_checker);
}
```

And the optimizer would `throw` after its `reset()`. 

So if we fail 1 time, it resets and throws. The throw is caught and we attempt to do it again. This time if it fails, since we're in the `catch` condition, then it will escalate to the controller server to pick up on. 

The only thing to add to that is to make sure that this exception is from the no trajectories condition or that we know for a fact that for any issue this is fine and we have `reset()` from it internally to the optimizer.

Reply:
add fallback behavior